File: ambience/ambictlnet/ambictl/Ambience/ServiceInterface.cs
using System.Collections.Generic;

namespace Ambience
{
    public class ServiceInterface
    {
        private readonly LidlModule _module;
        private readonly string _fullServiceName;

        public ServiceInterface(LidlModule mod, string name)
        {
            _module = mod;
            _fullServiceName = name;
        }

        public string AbsoluteName()
        {
            return _fullServiceName;
        }

        public string SyncServerName()
        {
            return $"{AbsoluteName()}::sync_server";
        }

        public string AsyncServerName()
        {
            return $"{AbsoluteName()}::async_server";
        }

        public string GetInclude()
        {
            return $"{System.IO.Path.GetFileNameWithoutExtension(_module.FileName())}_generated.hpp";
        }

        public Service Implement(string name, bool sync, bool external,
            Dictionary<string, ServiceInterface>? dependencies = null, string? cmakeTarget = null)
        {
            return new Service(name, this, sync, external, dependencies, cmakeTarget);
        }
    }
}

File: ambience/ambictl/ambictl/service_interface.py
from __future__ import annotations
from typing import Dict

import os

import ambictl.lidl_module
import ambictl.service

class ServiceInterface:
    module: LidlModule
    full_serv_name: str

    def __init__(self, mod: LidlModule, full_name: str) -> None:
        self.module = mod
        self.full_serv_name = full_name

    def absolute_name(self):
        return self.full_serv_name

    def sync_server_name(self):
        return f"{self.absolute_name()}::sync_server"

    def async_server_name(self):
        return f"{self.absolute_name()}::async_server"

    def get_include(self):
        return os.path.splitext(self.module.file_name())[0] + "_generated.hpp"

    def implement(self, name: str, *, sync: bool, extern: bool,
                  deps: Dict[str, ServiceInterface] = {}, cmake_target: str = "") -> Service:
        return Service(name, cmake_target, self, sync, extern, deps)

